File: loader/src/main/java/de/timolia/lactea/loader/module/discovery/DiscoveryIndex.java
package de.timolia.lactea.loader.module.discovery;

import com.google.common.collect.HashMultimap;
import com.google.common.collect.Multimap;
import java.io.DataInputStream;
import java.io.IOException;
import java.lang.annotation.Annotation;
import java.util.Collection;
import java.util.Enumeration;
import java.util.jar.JarEntry;
import java.util.jar.JarFile;
import javassist.bytecode.AnnotationsAttribute;
import javassist.bytecode.ClassFile;

/**
 * @author David (_Esel)
 */
public class DiscoveryIndex {
    private final Multimap<String, DiscoveryClass> index = HashMultimap.create();

    void addToIndex(DiscoveryClass discoveryClass) {
        for (javassist.bytecode.annotation.Annotation annotation : discoveryClass.getAnnotations()) {
            index.put(annotation.getTypeName(), discoveryClass);
        }
    }

    public void indexJarFile(JarFile jar) throws IOException {
        Enumeration<JarEntry> entries = jar.entries();
        while (entries.hasMoreElements()) {
            JarEntry jarEntry = entries.nextElement();
            if (jarEntry.getName().endsWith(".class")) {
                ClassFile classFile = new ClassFile(new DataInputStream(jar.getInputStream(jarEntry)));
                AnnotationsAttribute visible = (AnnotationsAttribute) classFile.getAttribute("RuntimeVisibleAnnotations");
                if (visible != null) {
                    addToIndex(new DiscoveryClass(classFile.getName(), visible.getAnnotations()));
                }
            }
        }
    }


    public Collection<DiscoveryClass> runDiscovery(String search) {
        return index.get(search);
    }

    public Collection<DiscoveryClass> runDiscovery(Class<? extends Annotation> search) {
        return runDiscovery(search.getName());
    }
}

File: core/src/main/kotlin/de/timolia/lactea/core/discovery/DiscoveryIndex.kt
package de.timolia.lactea.core.discovery

import com.google.common.collect.HashMultimap
import com.google.common.collect.Multimap
import javassist.bytecode.AnnotationsAttribute
import javassist.bytecode.ClassFile
import java.io.DataInputStream
import java.io.IOException
import java.util.*
import java.util.jar.JarEntry
import java.util.jar.JarFile

class DiscoveryIndex {
    private val index: Multimap<String, DiscoveryClass> = HashMultimap.create()

    fun addToIndex(discoveryClass: DiscoveryClass) {
        for (annotation in discoveryClass.annotations) {
            index.put(annotation.typeName, discoveryClass)
        }
    }

    @Throws(IOException::class)
    fun indexJarFile(jar: JarFile) {
        val entries: Enumeration<JarEntry> = jar.entries()
        while (entries.hasMoreElements()) {
            val jarEntry: JarEntry = entries.nextElement()
            if (jarEntry.name.endsWith(".class")) {
                val classFile = ClassFile(DataInputStream(jar.getInputStream(jarEntry)))
                val visible: AnnotationsAttribute = classFile.getAttribute("RuntimeVisibleAnnotations") as AnnotationsAttribute
                if (visible != null) {
                    addToIndex(DiscoveryClass(classFile.name, visible.annotations))
                }
            }
        }
    }

    fun runDiscovery(search: String) = DiscoveryResult(index[search])

    fun runDiscovery(search: Class<out Annotation>) = runDiscovery(search.name)
}
